src/qiskit_qec/circuits/css_code.py

```diff
@@ -0,0 +1,465 @@
+        if self._phenom:
+            states += ["0n", "1n"]
+        for state in states:
+            qc = QuantumCircuit()
+            qregs = []
+            qregs.append(QuantumRegister(code.n, name="code qubits"))
+            qregs.append(QuantumRegister(len(self.z_gauges), name="z auxs"))
+            qregs.append(QuantumRegister(len(self.x_gauges), name="x auxs"))
+            for qreg in qregs:
+                qc.add_register(qreg)
+            # prepare initial state
+            if state[0] == "1":
+                if basis == "z":
+                    qc.x(self.logical_x[0])
+                else:
+                    qc.x(self.logical_z[0])
+            if basis == "x":
+                qc.h(qregs[0])
+            # peform syndrome measurements
+            for t in range(T):
+                if state[-1] == "n" and self._phenom:
+                    for q in qregs[0]:
+                        qc.append(self._depol_error, [q])
+                # gauge measurements
+                if round_schedule == "zx":
+                    self._z_gauge_measurements(qc, t, state)
+                    self._x_gauge_measurements(qc, t, state)
+                elif round_schedule == "xz":
+                    self._x_gauge_measurements(qc, t, state)
+                    self._z_gauge_measurements(qc, t, state)
+                else:
+                    print("Round schedule " + round_schedule + " not supported.")
+            # final readout
+            creg = ClassicalRegister(code.n, name="final_readout")
+            qc.add_register(creg)
+            if basis == "x":
+                qc.h(qregs[0])
+            if state[-1] == "n" and self._phenom:
+                for q in qregs[0]:
+                    qc.append(self._meas_error, [q])
+            qc.measure(qregs[0], creg)
+            circuit[state] = qc
+
+        self.circuit = {}
+        self.noisy_circuit = {}
+        for state, qc in circuit.items():
+            if state[-1] == "n" and self._phenom:
+                self.noisy_circuit[state[0]] = qc
+            else:
+                self.circuit[state] = qc
+        if noise_model and not self._phenom:
+            for state, qc in circuit.items():
+                self.noisy_circuit[state] = noisify_circuit(qc, noise_model)
+
+        self._gauges4stabilizers = []
+        self._stabilizers = [self.x_stabilizers, self.z_stabilizers]
+        self._gauges = [self.x_gauges, self.z_gauges]
+        for j in range(2):
+            self._gauges4stabilizers.append([])
+            for stabilizer in self._stabilizers[j]:
+                gauges = []
+                for g, gauge in enumerate(self._gauges[j]):
+                    if set(stabilizer).intersection(set(gauge)) == set(gauge):
+                        gauges.append(g)
+                self._gauges4stabilizers[j].append(gauges)
+
+    def _get_code_properties(self):
+        if isinstance(self.code, StabSubSystemCode):
+            is_css = True
```

Comment:
How do you know that the given code is a CSS code? You should test to see if the given code is a CSS code and if not throw an error. To test if the code is CSS you can use the following result from Ted Yoder (which is in a paper that Andrew and I are just finishing up): It states the following:

Lemma 12.4 (Ted Yoder’s lemma). Let G = [G_X |G_Z ] be the generator matrix of stabilizer group.
The group is strictly CSS if, and only if, rank(G_X ) + rank(G_Z ) = n − k.

So, implement this simple test and you are done.

Reply:
We'll open a separate issue for this.